### PR TITLE
APM-311115: Added URL and API token permissions validation

### PR DESCRIPTION
**Pull request by mswiatkowska**

### CODE REVIEW

File: scripts/deploy-helm.sh [hunk 1]
```diff
@@ -30,9 +30,42 @@ check_if_parameter_is_empty()
...
+  URL=$1
+  if RESPONSE=$(curl -k -s -X POST -d "{\"token\":\"$DYNATRACE_ACCESS_KEY\"}" "$URL/api/v2/apiTokens/lookup" -w "<<HTTP_CODE>>%{http_code}" -H "accept: application/json; charset=utf-8" -H "Content-Type: application/json; charset=utf-8" -H "Authorization: Api-Token $DYNATRACE_ACCESS_KEY"); then
+    CODE=$(sed -rn 's/.*<<HTTP_CODE>>(.*)$/\1/p' <<<"$RESPONSE")
+    RESPONSE=$(sed -r 's/(.*)<<HTTP_CODE>>.*$/\1/' <<<"$RESPONSE")
+    if [ "$CODE" -ge 300 ]; then
+      echo "Failed to check Dynatrace API token permissions - please verify provided URL and API token. $RESPONSE"
+      exit 1
+    fi
+    for scope in "${API_TOKEN_SCOPES[@]}"; do
+      if ! grep -q "$scope" <<< "$RESPONSE"; then
```
Comment (pgutkowski) [inline scripts/deploy-helm.sh:44]: I'm wondering if we should wrap scope with another quotes to match exact JSON token `"scope"` and avoid false positives?

File: scripts/deploy-helm.sh [hunk 2]
```diff
@@ -30,9 +30,42 @@ check_if_parameter_is_empty()
...
+        echo "Missing permission for the API token: $scope."
+        echo "Please enable all required permissions: ${API_TOKEN_SCOPES[*]} for chosen deployment type: $DEPLOYMENT_TYPE"
+        exit 1
+      fi
+    done
+  else
+      echo "Failed to check Dynatrace API token permissions - please verify provided URL and API token."
```
Comment (sq2gxo) [inline scripts/deploy-helm.sh:51]: - we could be more specific in message, like "please verify provided ingest url: ${URL}"
- is it the case for connectivity error? then we should not mention TOKEN in this message

File: scripts/deploy-helm.sh [hunk 3]
```diff
@@ -30,9 +30,42 @@ check_if_parameter_is_empty()
...
-print_help()
+check_api_token()
 {
-   printf "
+  URL=$1
+  if RESPONSE=$(curl -k -s -X POST -d "{\"token\":\"$DYNATRACE_ACCESS_KEY\"}" "$URL/api/v2/apiTokens/lookup" -w "<<HTTP_CODE>>%{http_code}" -H "accept: application/json; charset=utf-8" -H "Content-Type: application/json; charset=utf-8" -H "Authorization: Api-Token $DYNATRACE_ACCESS_KEY"); then
+    CODE=$(sed -rn 's/.*<<HTTP_CODE>>(.*)$/\1/p' <<<"$RESPONSE")
+    RESPONSE=$(sed -r 's/(.*)<<HTTP_CODE>>.*$/\1/' <<<"$RESPONSE")
+    if [ "$CODE" -ge 300 ]; then
+      echo "Failed to check Dynatrace API token permissions - please verify provided URL and API token. $RESPONSE"
```
Comment (sq2gxo) [inline scripts/deploy-helm.sh:40]: please add value of URL in log message - it will help customer to investigate the problem

**Review by pgutkowski — DISMISSED** [review]

**Review by pgutkowski — APPROVED** [review]

**Review by Dawid-Kaszubski — DISMISSED** [review]

**Review by dtwwolinski — APPROVED** [review]

**Review by sq2gxo — APPROVED** [review]

**Review by karolkieglerski — DISMISSED** [review]

File: src/lib/credentials.py [hunk 4]
```diff
@@ -51,7 +51,7 @@ def get_dynatrace_log_ingest_url_from_env():
     url = os.environ.get(_DYNATRACE_LOG_INGEST_URL_SECRET_NAME, None)
     if url is None:
         raise Exception("{env_var} environment variable is not set".format(env_var=_DYNATRACE_LOG_INGEST_URL_SECRET_NAME))
-    return url
+    return url.rstrip('/')
```
Comment (sq2gxo) [inline src/lib/credentials.py:54]: also need to update ARM template and handle '/' at the end

**Review by sq2gxo — APPROVED** [review]

**Review by sq2gxo — APPROVED** [review]

**Review by Dawid-Kaszubski — APPROVED** [review]

**Pull request merged**